File: cpp/problem507.cpp
#include <mystd.h>
using namespace MYSTD;
using namespace std;

/**
 * follow the definition of perfect number
 * Here we consider all numbers below sqrt(num) which are possible divisors.
 */
class Solution {
public:
    bool checkPerfectNumber(int num) {
        if( num <= 1 ) return false;
        
        int sumOfDivisor = 0;
        int t = sqrt(num + 0.9);

        // For perfect square number, the square root should be added only once
        if( t*t == num ){ sumOfDivisor += t; --t; } 
        // 1 should be added, while num itself should not.
        sumOfDivisor += 1;                           
        // other divisors
        for(int i=2; i<=t; ++i){
            if( num%i == 0 ){
                sumOfDivisor += i + num/i;
            } 
        }
        return sumOfDivisor == num;
    }
};

int main(int argc, char const *argv[]){
    Solution sol;
    // test case in Leetcode Problem
    cout << sol.checkPerfectNumber(28) << endl;
    // Other extreme case
    cout << sol.checkPerfectNumber(0) << endl;
    cout << sol.checkPerfectNumber(1) << endl;
    cout << sol.checkPerfectNumber(25) << endl;
    return 0;
}

File: python/problem507.py
import math
class Solution:
    def checkPerfectNumber(self, num: int) -> bool:
        if num == 1:
            return False
        n = int(math.sqrt(abs(num))) + 1
        li = [1]
        for i in range(2,n):
            
            if num%i == 0:
                li.append(num//i)
                li.append(i)
        
        return num == sum(li)


import math
class Solution:
    def checkPerfectNumber(self, num: int) -> bool:
        if num <= 1:
            return False
        n = int(math.sqrt(num)) + 1
        li = 1
        for i in range(2,n):
            
            if num%i == 0:
                li += i + num//i
        return num == li
              